File: LeetCode/二分查找/34.在排序数组中查找元素的第一个和最后一个位置.cpp
/*
 * @lc app=leetcode.cn id=34 lang=cpp
 *
 * [34] 在排序数组中查找元素的第一个和最后一个位置
 */

// @lc code=start
#include<bits/stdc++.h>
using namespace std;

class Solution {
public:
    vector<int> searchRange(vector<int>& nums, int target) {
        int l = 0, r = nums.size() - 1;
        while(l < r) {
            int mid = l + r >> 1;
            cout<<"idx"<<mid<<"mid"<<nums[mid]<<endl;
            if(nums[mid] == target){
                int s = -1, e =-1;
                for(int j = l; j <= r; j++){ 
                    if(nums[j] == target && s == -1) 
                        s = j;
                    if(nums[j] == target) 
                        e = j;
                //     if(nums[j] != target && s != -1 && e == -1)
                //         e = j;
                }
                cout<<s<<" "<<e<<endl;
                return {s,e};
            }
            else if(nums[mid] < target)
                l = mid + 1;
            else
                r = mid;
        }
        if(nums.size() != 0 &&nums[l] == target)
            return {l,l};
        return {-1,-1};
    }
};
// @lc code=end

int main()
{
    Solution solve;
    vector<int> nums = {5,7,7,8,8,10};
    vector<int> res = solve.searchRange(nums, 8);
    cout<<res[0]<<" "<<res[1]<<endl;
    return 0;
}

File: LeetCode/二分查找/34.在排序数组中查找元素的第一个和最后一个位置.py
from typing import List


class Solution:
    def lower1(self, nums, target: int):
        # 闭区间 [left, right]
        l = 0
        r = len(nums) - 1
        while l <= r : # 区间不为空
            m = (l + r) // 2
            if nums[m] < target:
                l = m + 1
            else:
                r = m - 1
        return l
    
    def lower2(self, nums, target: int):
        # 左闭右开区间 [left, right]
        l = 0
        r = len(nums)
        while l < r : # 区间不为空
            m = (l + r) // 2
            if nums[m] < target:
                l = m + 1 # [mid+1, r)
            else:
                r = m  # [l, mid)
        return l
    
    def lower3(self, nums, target: int):
        # 左闭右开区间 (left, right)
        l = -1
        r = len(nums) 
        while l + 1 < r : # 区间不为空
            m = (l + r) // 2
            if nums[m] < target:
                l = m # (mid+1, r)
            else:
                r = m  # [l, mid)
        return l   
    
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        start = self.lower1(nums, target)
        if start == len(nums) or nums[start] != target:
            return [-1, -1]
        end = self.lower1(nums, target+1) - 1
        return [start, end]
    
nums = [5,7,7,8,8,10]
s = Solution()
print(s.searchRange(nums, 8))